File: rust/src/update_source.rs
//! Build `skills add` source arguments for updates (port of update-source.ts).

use crate::urlutil;
use serde_json::Value;

#[derive(Clone, Debug, Default)]
pub struct UpdateSourceEntry {
    pub source: String,
    pub source_url: Option<String>,
    pub source_type: Option<String>,
    pub r#ref: Option<String>,
    pub skill_path: Option<String>,
}

fn truthy(v: Option<&Value>) -> Option<String> {
    v.and_then(|x| x.as_str())
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
}

impl UpdateSourceEntry {
    /// Build from a lock-file JSON entry (empty strings count as absent).
    pub fn from_json(e: &Value) -> Self {
        UpdateSourceEntry {
            source: e
                .get("source")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            source_url: truthy(e.get("sourceUrl")),
            source_type: truthy(e.get("sourceType")),
            r#ref: truthy(e.get("ref")),
            skill_path: truthy(e.get("skillPath")),
        }
    }
}

pub fn format_source_input(source_url: &str, r#ref: Option<&str>) -> String {
    match r#ref.filter(|r| !r.is_empty()) {
        None => source_url.to_string(),
        Some(r) => format!("{}#{}", source_url, r),
    }
}

fn derive_skill_folder(skill_path: &str) -> String {
    let mut folder = skill_path.to_string();
    if folder.ends_with("/SKILL.md") {
        folder.truncate(folder.len() - 9);
    } else if folder.ends_with("SKILL.md") {
        folder.truncate(folder.len() - 8);
    }
    if folder.ends_with('/') {
        folder.pop();
    }
    folder
}

fn supports_appended_subpath(source: &str) -> bool {
    if source.starts_with("git@") || source.starts_with("ssh://") || source.ends_with(".git") {
        return false;
    }
    if source.starts_with("http://") || source.starts_with("https://") {
        return match urlutil::parse(source) {
            Some(u) => {
                let h = urlutil::hostname(&u);
                h == "github.com" || h == "gitlab.com"
            }
            None => false,
        };
    }
    true
}

fn is_bare_shorthand(source: &str) -> bool {
    !source.contains(':') && !source.starts_with('.') && !source.starts_with('/')
}

fn get_local_source(e: &UpdateSourceEntry) -> Option<String> {
    if let Some(u) = &e.source_url {
        return Some(u.clone());
    }
    let requires_url = matches!(e.source_type.as_deref(), Some("git") | Some("gitlab"));
    if requires_url && is_bare_shorthand(&e.source) {
        return None;
    }
    Some(e.source.clone())
}

/// Cloneable repository URL for project update checks.
pub fn build_local_clone_source(e: &UpdateSourceEntry) -> Option<String> {
    let source = get_local_source(e)?;
    if e.source_type.as_deref() == Some("github") && is_bare_shorthand(&source) {
        return Some(format!(
            "https://github.com/{}.git",
            source.strip_suffix(".git").unwrap_or(&source)
        ));
    }
    Some(source)
}

pub fn should_use_full_depth_for_update(e: &UpdateSourceEntry) -> bool {
    if e.skill_path.is_none() {
        return false;
    }
    let source = match e.source_type.as_deref() {
        Some(t) if t != "github" => get_local_source(e),
        _ => Some(e.source.clone()),
    };
    source
        .map(|s| !supports_appended_subpath(&s))
        .unwrap_or(false)
}

fn append_folder_and_ref(source: &str, skill_path: &str, r#ref: Option<&str>) -> String {
    if !supports_appended_subpath(source) {
        return format_source_input(source, r#ref);
    }
    let folder = derive_skill_folder(skill_path);
    let with_folder = if folder.is_empty() {
        source.to_string()
    } else {
        format!("{}/{}", source, folder)
    };
    format_source_input(&with_folder, r#ref)
}

/// Source argument for `skills add` during a global update.
pub fn build_update_install_source(e: &UpdateSourceEntry) -> Option<String> {
    let non_github = matches!(e.source_type.as_deref(), Some(t) if t != "github");
    match &e.skill_path {
        None => {
            let source = if non_github {
                get_local_source(e)
            } else {
                Some(e.source_url.clone().unwrap_or_else(|| e.source.clone()))
                    .filter(|s| !s.is_empty())
            };
            Some(format_source_input(&source?, e.r#ref.as_deref()))
        }
        Some(sp) => {
            let source = if non_github {
                get_local_source(e)
            } else {
                Some(e.source.clone())
            };
            let source = source.filter(|s| !s.is_empty())?;
            Some(append_folder_and_ref(&source, sp, e.r#ref.as_deref()))
        }
    }
}

/// Source argument for `skills add` during a project update.
pub fn build_local_update_source(e: &UpdateSourceEntry) -> Option<String> {
    let source = get_local_source(e)?;
    match &e.skill_path {
        None => Some(format_source_input(&source, e.r#ref.as_deref())),
        Some(sp) => Some(append_folder_and_ref(&source, sp, e.r#ref.as_deref())),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn entry(
        source: &str,
        url: Option<&str>,
        t: &str,
        r: Option<&str>,
        sp: Option<&str>,
    ) -> UpdateSourceEntry {
        UpdateSourceEntry {
            source: source.into(),
            source_url: url.map(|s| s.into()),
            source_type: Some(t.into()),
            r#ref: r.map(|s| s.into()),
            skill_path: sp.map(|s| s.into()),
        }
    }

    #[test]
    fn github_shorthand_with_folder_and_ref() {
        let e = entry(
            "o/r",
            Some("https://github.com/o/r.git"),
            "github",
            Some("v1"),
            Some("skills/x/SKILL.md"),
        );
        assert_eq!(
            build_update_install_source(&e).as_deref(),
            Some("o/r/skills/x#v1")
        );
        assert!(!should_use_full_depth_for_update(&e));
        let root = entry("o/r", None, "github", None, Some("SKILL.md"));
        assert_eq!(build_update_install_source(&root).as_deref(), Some("o/r"));
    }

    #[test]
    fn generic_git_sources() {
        let e = entry(
            "team/repo",
            Some("https://git.example.com/team/repo.git"),
            "git",
            None,
            Some("skills/x/SKILL.md"),
        );
        assert_eq!(
            build_local_update_source(&e).as_deref(),
            Some("https://git.example.com/team/repo.git")
        );
        assert!(should_use_full_depth_for_update(&e));
        let legacy = entry("team/repo", None, "git", None, None);
        assert_eq!(build_local_update_source(&legacy), None);
        assert_eq!(build_local_clone_source(&legacy), None);
        let gh = entry("o/r", None, "github", None, None);
        assert_eq!(
            build_local_clone_source(&gh).as_deref(),
            Some("https://github.com/o/r.git")
        );
        let ssh = entry(
            "git@github.com:o/r.git",
            None,
            "git",
            Some("main"),
            Some("a/SKILL.md"),
        );
        assert_eq!(
            build_local_update_source(&ssh).as_deref(),
            Some("git@github.com:o/r.git#main")
        );
    }
}

File: dotnet/lib/Devolutions.AgentSkills/UpdateSource.cs
// Build `skills add` source arguments for updates (port of update-source.ts).

using System.Text.Json.Nodes;

namespace Skills;

internal sealed record UpdateSourceEntry(string Source, string? SourceUrl = null, string? SourceType = null, string? Ref = null, string? SkillPath = null)
{
    /// Build from a lock-file JSON entry (empty strings count as absent).
    public static UpdateSourceEntry FromJson(JsonNode? e) =>
        new(Json.Str(e, "source") ?? "", Json.NonEmpty(e, "sourceUrl"), Json.NonEmpty(e, "sourceType"), Json.NonEmpty(e, "ref"), Json.NonEmpty(e, "skillPath"));
}

internal static class UpdateSource
{
    public static string FormatSourceInput(string sourceUrl, string? r) => string.IsNullOrEmpty(r) ? sourceUrl : $"{sourceUrl}#{r}";

    private static string DeriveSkillFolder(string skillPath)
    {
        var folder = skillPath;
        if (folder.EndsWith("/SKILL.md")) folder = folder[..^9];
        else if (folder.EndsWith("SKILL.md")) folder = folder[..^8];
        if (folder.EndsWith('/')) folder = folder[..^1];
        return folder;
    }

    private static bool SupportsAppendedSubpath(string source)
    {
        if (source.StartsWith("git@") || source.StartsWith("ssh://") || source.EndsWith(".git")) return false;
        if (source.StartsWith("http://") || source.StartsWith("https://"))
            return WebUrl.Parse(source) is { Hostname: "github.com" or "gitlab.com" };
        return true;
    }

    private static bool IsBareShorthand(string source) => !source.Contains(':') && !source.StartsWith('.') && !source.StartsWith('/');

    private static string? GetLocalSource(UpdateSourceEntry e)
    {
        if (e.SourceUrl != null) return e.SourceUrl;
        var requiresUrl = e.SourceType is "git" or "gitlab";
        return requiresUrl && IsBareShorthand(e.Source) ? null : e.Source;
    }

    /// Cloneable repository URL for project update checks.
    public static string? BuildLocalCloneSource(UpdateSourceEntry e)
    {
        var source = GetLocalSource(e);
        if (source == null) return null;
        if (e.SourceType == "github" && IsBareShorthand(source))
            return $"https://github.com/{(source.EndsWith(".git") ? source[..^4] : source)}.git";
        return source;
    }

    public static bool ShouldUseFullDepthForUpdate(UpdateSourceEntry e)
    {
        if (e.SkillPath == null) return false;
        var source = e.SourceType is { } t && t != "github" ? GetLocalSource(e) : e.Source;
        return source != null && !SupportsAppendedSubpath(source);
    }

    private static string AppendFolderAndRef(string source, string skillPath, string? r)
    {
        if (!SupportsAppendedSubpath(source)) return FormatSourceInput(source, r);
        var folder = DeriveSkillFolder(skillPath);
        return FormatSourceInput(folder.Length == 0 ? source : $"{source}/{folder}", r);
    }

    /// Source argument for `skills add` during a global update.
    public static string? BuildUpdateInstallSource(UpdateSourceEntry e)
    {
        var nonGithub = e.SourceType is { } t && t != "github";
        if (e.SkillPath is not { } sp)
        {
            var source = nonGithub ? GetLocalSource(e) : e.SourceUrl ?? e.Source;
            return string.IsNullOrEmpty(source) ? null : FormatSourceInput(source, e.Ref);
        }
        var s = nonGithub ? GetLocalSource(e) : e.Source;
        return string.IsNullOrEmpty(s) ? null : AppendFolderAndRef(s, sp, e.Ref);
    }

    /// Source argument for `skills add` during a project update.
    public static string? BuildLocalUpdateSource(UpdateSourceEntry e)
    {
        var source = GetLocalSource(e);
        if (source == null) return null;
        return e.SkillPath is { } sp ? AppendFolderAndRef(source, sp, e.Ref) : FormatSourceInput(source, e.Ref);
    }
}

internal sealed record DiscoveredSkillLocation(string Name, string SkillPath);

internal sealed record SkillLocationResolution(List<string> DeletedSkills, List<string> AmbiguousSkills, Dictionary<string, string> ResolvedPaths);

/// Resolve locked skills against their current locations (port of skill-relocation.ts).
internal static class SkillRelocation
{
    private static string NormalizePath(string p)
    {
        var s = p.Replace('\\', '/');
        var sb = new System.Text.StringBuilder();
        foreach (var c in s)
        {
            if (c == '/' && sb.Length > 0 && sb[^1] == '/') continue;
            sb.Append(c);
        }
        return sb.ToString();
    }

    /// Exact paths win; a missing path is a relocation only when exactly one
    /// discovered skill has the same normalized name. Ambiguity fails closed.
    public static SkillLocationResolution Resolve(IEnumerable<string> locked, JsonObject lockSkills, IReadOnlyList<DiscoveredSkillLocation> discovered)
    {
        var discoveredPaths = discovered.Select(d => NormalizePath(d.SkillPath)).ToHashSet();
        var byName = new Dictionary<string, List<string>>();
        foreach (var d in discovered)
        {
            var key = SkillDiscovery.NormalizeSkillName(d.Name);
            if (!byName.TryGetValue(key, out var list)) byName[key] = list = [];
            var p = NormalizePath(d.SkillPath);
            if (!list.Contains(p)) list.Add(p);
        }
        var output = new SkillLocationResolution([], [], []);
        foreach (var name in locked)
        {
            if (Json.NonEmpty(lockSkills[name], "skillPath") is not { } lockedPath) continue;
            var nlp = NormalizePath(lockedPath);
            var candidates = byName.GetValueOrDefault(SkillDiscovery.NormalizeSkillName(name)) ?? [];
            if (candidates.Count > 1)
            {
                output.AmbiguousSkills.Add(name);
                continue;
            }
            if (discoveredPaths.Contains(nlp)) output.ResolvedPaths[name] = nlp;
            else if (candidates.Count == 1) output.ResolvedPaths[name] = candidates[0];
            else output.DeletedSkills.Add(name);
        }
        return output;
    }
}
